bundles/org.openhab.voice.voicerss/src/main/java/org/openhab/voice/voicerss/tool/CreateTTSCache.java

```diff
@@ -106,7 +106,7 @@ private void generateCacheForMessage(String apiKey, String cacheDir, String loca
             return;
         }
         CachedVoiceRSSCloudImpl impl = new CachedVoiceRSSCloudImpl(cacheDir);
-        File cachedFile = impl.getTextToSpeechAsFile(apiKey, trimmedMsg, locale, voice, "MP3");
+        File cachedFile = impl.getTextToSpeechAsFile(apiKey, trimmedMsg, locale, voice, "MP3", null);
```

Comment:
I am a little lost. Why is MP3 hardcoded here ?
And using null as last new parameter can not lead to problems (NPE) ?

more generally, now, using VoiceRSS, what will be with your changes the default codec used ? Will it be MP3 if the sink supports MP3 ?

Reply:
> I am a little lost. Why is MP3 hardcoded here ?

"MP3" was already hardcoded before and not parameterized in the commit to "add support for OGG and AAC audio formats" because it's probably unused. So I also didn't touch it.

This file is meant to be called from the CLI, see [Caching in README.md](https://github.com/openhab/openhab-addons/tree/main/bundles/org.openhab.voice.voicerss#caching). Perhaps it should be deleted as calling the [`openhab:voice say <text>` command](https://www.openhab.org/docs/administration/runtime.html#voice) in the runtime console has the same effect.

> And using null as last new parameter can not lead to problems (NPE) ?

No risk of a NPE: ` if (!Objects.equals(format, "44khz_16bit_mono")) { ... }`

> More generally, using VoiceRSS with your changes, what will be the default codec used ? Will it be MP3 if the sink supports MP3?

Hm. Good question. I think so, if the sink does not support WAV. That's decided in [`VoiceManagerImpl`](https://github.com/openhab/openhab-core/blob/0a778df0adddd90f3db219662c1d7c7369bed14f/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L217):

```
Set<AudioFormat> audioFormats = tts.getSupportedFormats();
...
AudioFormat audioFormat = getBestMatch(audioFormats, sink.getSupportedFormats());
```

So I think that's not really an implementation concern of the VoiceRSS binding but the openhab-core classes.

Reply:
I see that it was previously hardcoded but I did not remember why. As I was probably the one who added OGG/AAC, if I left MP3 hardcoded here, there is certainly a good reason, I hope :) But I will check that again, considering your comments about the CLI command to build the cache.

I would like to be sure that these changes will not lead to users "loosing" their cache in practice because now the binding is now using another codec and other files in the cache.
Currently, I see that my cached files have all the MP3 extension.

Let me the time to analyse more in details the code and to test your changes.

Reply:
I made sure not to change the filename for previous uses of the default 44khz_16bit_mono format:

```
/**
 * Gets a unique filename [...] suffixed by the format if it is not the default of "44khz_16bit_mono".
 */
private String getUniqueFilenameForText(String text, String locale, String voice, String format) {
    ...
    if (!Objects.equals(format, "44khz_16bit_mono")) {
        filename += "_" + format;
    }
    ...
}
```

Thanks for your review so far.

Reply:
A lot of audio sinks support MP3 and WAV: javasound, webaudio, chromecast, kodi, squeezebox, ...

`VoiceManagerImpl#getPreferredFormat` seems to prefer WAV, so I guess you're right that this change might lead to cache misses.

But it would be somewhat unfitting to have special "prefer MP3" code only in this binding. Better to implement something in openhab-core.

There's code duplication between `VoiceManagerImpl#getPreferredFormat` and `AudioFormat#getPreferredFormat`, the latter has a strange `defaultFrequency = 16384` setting, that is not contained in the [list of common sampling frequencies in Wikipedia](https://en.m.wikipedia.org/wiki/Sampling_(signal_processing)) (the correct value would be 16000 Hz). This is corrected in `VoiceManagerImpl` to 44100 Hz. I already wrote that bitRate is practically unused. bigEndian only makes sense for WAV with a bit depth of 16 bit but not for 8 bit WAV or MP3 or other compressed formats. Maybe it's time for a little cleanup and refactoring there.

Reply:
The duplicated code is something I have detected several years ago, I mentioned it somewhere in an issue but it remained like that as I got no answer.

Reply:
In case, WAV iis preferred with your changes, it would impact mainly users who built the cache with predefined texts. I don't think this is something very common. In this case, this should be mentioned in the release notes as a breaking change (but I assume this will impact really a small number of users).

Reply:
Ok this class concerns only the external tool to create cached files. 
It could be certainly improved to support other audio format through a new command line argument, probably in a separate PR. 

What remains important to understand and test is what will become the new default format after your changes.

Reply:
If I change the default audio sink to "webaudio" and call `voice say This is test` than a .wav file is created in the cache folder.

How can I mention this change in the release notes? Can this be part of this pull request?

Reply:
Not in this PR as the file to be changed in another repo. 

No way to give to MP3 a higher priority so that we keep a full backward compatibility?

Reply:
I will test with Sonos as renderer.

Reply:
Maybe we should just add an advanced parameter to let the user force an audio format. 
By default, the binding will support WAV, NP3  OGG, ... If set to MP3 for example, the binding will only consider MP3 format. 
Like that, er have a way to keep backward  compatibility + compatibility with the cache tool. 
WDYT?

Reply:
A config option to opt-in to WAV support could be implemented. The return value of `getSupportedFormats()` could be without WAV in the default case then. But I think this is not something a single binding should do as a special case.

The code that decides what audio format to use is in openhab-core. I can't say if a sink can prefer a format by returning it first in the of list of supported formats. Oh, I see that it's a `Set` and not a list, so I guess returning a `LinkedHashSet`  would be possible but isn't really part of the contract of `AudioSink`.

Another possibility is to mention how to batch convert the cache in the release notes:
```
for i in userdata/voicerss/cache/*.mp3; do sox "$i" "${i%.mp3}.wav"; done
```

Reply:
> By default, the binding will support WAV, MP3 , OGG, ... If set to MP3 for example, the binding will only consider MP3 format.
> Like that, we have a way to keep backward compatibility

Listing the preferred audio formats to use as a config option in openhab-core would be a good idea. But would not solve our problem here, because -core preferred wav and should continue to do so when such an option is introduced, I guess.

So if you do not think a remark in the release notes is enough we should start to implement an option for the voicerss binding and maybe later on do something about it in -core. Perhaps even changing `AudioSink` to return a `List`. A sink should have the possibility to specify a preference.

Reply:
What I was thinking of is clearly an option in voicerss. My concern is users who created cached files for voicerss using the provided tool. Currently all these files are MP3 files. 

My idea was a binding parameter that would limit the audio formats handled by the binding to a particular audio format, MP3 as an example. 

But now I think the best option is just to enhance the tool to let the user select the output audio format. Like that, to switch to WAV, files, the user will have only to rerurn his command but asking WAV format. 

Can you enhance the tool with one additional command line argument that will be the audio codec to be used?

Reply:
Okay, I will do that!